src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java

```diff
@@ -32,48 +37,90 @@ public RoleAssignmentsService(RoleService roleService) {
-                    } catch (Exception e) {
-                        val message = String.format("Failure while removing roles %1$s from user %2$s at caseload %3$s. Some roles may not have been removed. Continuing with next user.",specification.getRolesToRemove(), username, caseload);
-                        log.warn(message, e);
-                    }
-                } catch (Exception e) {
-                    val message = String.format("Failure while assigning roles %1$s to user %2$s. No roles will be removed from this user. Continuing with next user.", specification.getRolesToAssign(), username);
-                    log.warn(message, e);
-                }
-            });
-        }
+        return specification.getCaseloads().stream().map(caseload -> getResultsForCaseload(specification, caseload)).collect(toList());
     }
 
-    private void removeRolesFromUserAtCaseload(String caseload, String username, List<String> rolesToRemove) {
-        rolesToRemove.forEach(roleCodeToRemove -> {
-            try {
-                roleService.removeRole(username, caseload, roleCodeToRemove);
+    private RoleAssignmentStats getResultsForCaseload(@Valid RoleAssignmentsSpecification specification, String caseload) {
 
-            } catch (HttpClientErrorException.NotFound notFoundException) {
-                log.info("Username {} does not have role {} at caseload {}", username, roleCodeToRemove, caseload);
-            }
+        val usernamesForCaseload = findUsernamesMatchingRolesAtCaseload(specification.getRolesToMatch(), caseload);
+
+        log.info("Found {} users for the {} caseload: {}.", usernamesForCaseload.size(), caseload, usernamesForCaseload);
+        val results = RoleAssignmentStats.builder()
+                .numMatchedUsers(usernamesForCaseload.size())
+                .caseload(caseload)
+                .build();
+
+        usernamesForCaseload.forEach(username -> {
+            val assignResults = assignRolesToUser(username, specification.getRolesToAssign());
+            val unassigningResults = assignResults.containsKey(FAIL)
+                    ? Map.<RoleAssignmentStats.Status, Long>of()
+                    : removeRolesFromUserAtCaseload(caseload, username, specification.getRolesToRemove());
+            results.addAssignResults(assignResults);
+            results.addUnassignResults(unassigningResults);
         });
+
+        telemetryClient.trackEvent("UpdateRollAssignment", results.toMap(), null);
+        return results;
+    }
+
+    private Map<RoleAssignmentStats.Status, Long> assignRolesToUser(final String username, final List<String> rolesToAssign) {
+        return rolesToAssign.stream().reduce(new HashMap<RoleAssignmentStats.Status, Long>(), (results, role) -> {
+            if (!results.containsKey(FAIL)) {
+                val status = assignRole(username, role);
+                results.compute(status, (key, count) -> count == null ? 1 : count + 1);
```

Comment:
still seems very complicated with the `status`es.  Why can't you pass the `results` down and just individually increment as necessary rather than having to keep a map of all the successes / failures, pass it all back up and then loop over them all etc.

Reply:
Did it like this originally as:
 * Parameters should normally be treated as immutable
 * Not a big fan of modifying in-memory state of an object across multiple locations.
  
Switched it to the suggested approach..